src/synapse/api.py

```diff
@@ -0,0 +1,459 @@
+#!/usr/bin/env python3
+
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Helper module used to manage interactions with Synapse API."""
+
+# pylint: disable=too-few-public-methods, too-many-arguments
+
+import hashlib
+import hmac
+import logging
+import re
+import typing
+
+import requests
+from requests.adapters import HTTPAdapter
+from urllib3.util import Retry
+
+from charm_state import CharmState
+from constants import MJOLNIR_MANAGEMENT_ROOM, MJOLNIR_MEMBERSHIP_ROOM, SYNAPSE_URL
+from user import User
+
+logger = logging.getLogger(__name__)
+
+# The API version that should be used is described in the documentation:
+# https://matrix-org.github.io/synapse/latest/usage/administration/index.html
+ADD_USER_ROOM_URL = f"{SYNAPSE_URL}/_synapse/admin/v1/join"
+CREATE_ROOM_URL = f"{SYNAPSE_URL}/_matrix/client/v3/createRoom"
+DEACTIVATE_ACCOUNT_URL = f"{SYNAPSE_URL}/_synapse/admin/v1/deactivate"
+LIST_ROOMS_URL = f"{SYNAPSE_URL}/_synapse/admin/v1/rooms"
+LIST_USERS_URL = f"{SYNAPSE_URL}/_synapse/admin/v2/users?from=0&limit=10&name="
+LOGIN_URL = f"{SYNAPSE_URL}/_synapse/admin/v1/users"
+REGISTER_URL = f"{SYNAPSE_URL}/_synapse/admin/v1/register"
+SYNAPSE_VERSION_REGEX = r"(\d+\.\d+\.\d+(?:\w+)?)\s?"
+VERSION_URL = f"{SYNAPSE_URL}/_synapse/admin/v1/server_version"
+
+
+class APIError(Exception):
+    """Exception raised when something fails while calling the API.
+
+    Attrs:
+        msg (str): Explanation of the error.
+    """
+
+    def __init__(self, msg: str):
+        """Initialize a new instance of the APIError exception.
+
+        Args:
+            msg (str): Explanation of the error.
+        """
+        self.msg = msg
+
+
+class NetworkError(APIError):
+    """Exception raised when requesting API fails due network issues."""
+
+
+class GetNonceError(APIError):
+    """Exception raised when getting nonce fails."""
+
+
+class GetVersionError(APIError):
+    """Exception raised when getting version fails."""
+
+
+class VersionUnexpectedContentError(GetVersionError):
+    """Exception raised when output of getting version is unexpected."""
+
+
+class GetRoomIDError(APIError):
+    """Exception raised when getting room id fails."""
+
+
+class GetUserIDError(APIError):
+    """Exception raised when getting user id fails."""
+
+
+class UserExistsError(APIError):
+    """Exception raised when checking if user exists fails."""
+
+
+class GetAccessTokenError(APIError):
+    """Exception raised when getting access token fails."""
+
+
+class RegisterUserError(APIError):
+    """Exception raised when registering user fails."""
+
+
+# admin_access_token is not a password
+def register_user(
+    registration_shared_secret: str,
+    user: User,
+    server: typing.Optional[str] = None,
+    admin_access_token: typing.Optional[str] = None,  # nosec
```

Comment:
Is this `nosec` still needed since it seems that it is being defaulted to `None` now?

Reply:
Related: #52